Containerized Docker Application Lifecycle - 04.10 (#10614)
dotnet/docs: docs/standard/containerized-lifecycle-architecture/design-develop-containerized-apps/docker-apps-inner-loop-workflow.md

>
>For instructions on setting up Docker for Mac, go to <https://docs.docker.com/docker-for-mac/>.

In addition, you'll need a code editor so that you can actually develop your application while using Docker CLI.

Microsoft provides Visual Studio Code, which is a lightweight code editor that is supported on Mac, Windows, and Linux, and provides IntelliSense with [support for many languages](https://code.visualstudio.com/docs/languages/overview) (JavaScript, .NET, Go, Java, Ruby, Python, and most modern languages), [debugging](https://code.visualstudio.com/Docs/editor/debugging), [integration with Git](https://code.visualstudio.com/Docs/editor/versioncontrol) and [extensions support](https://code.visualstudio.com/docs/extensions/overview). This editor is a great fit for Mac and Linux developers. In Windows, you also can use the full Visual Studio application.

> [!INFORMATION]
>
> For instructions on installing Visual Studio Code for Windows, Mac, or Linux, go to <https://code.visualstudio.com/docs/setup/setup-overview/>.
>
> For instructions on setting up Docker for Mac, go to <https://docs.docker.com/docker-for-mac/>.

> You can work with Docker CLI and write your code using any code editor, but if you use Visual Studio Code with the Docker extension, it gets easy to author `Dockerfile` and `docker-compose.yml` files in your workspace. You can also run tasks and scripts from the Visual Studio Code IDE, to execute Docker commands using the Docker CLI underneath.

Reviewer: ```suggestion
You can work with Docker CLI and write your code using any code editor, but using Visual Studio Code with the Docker extension makes it easy to author `Dockerfile` and `docker-compose.yml` files in your workspace. You can also run tasks and scripts from the Visual Studio Code IDE to execute Docker commands using the Docker CLI underneath.
```

The Docker extension for VS Code provides the following features:

- Automatic `Dockerfile` and `docker-compose.yml` file generation

- Syntax highlighting and hover tips for `docker-compose.yml` and `Dockerfile` files

- IntelliSense (completions) for `Dockerfile` and `docker-compose.yml` files

- Linting (errors and warnings) for `Dockerfile` files

- Command Palette (F1) integration for the most common Docker commands

- Explorer integration for managing Images and Containers

- Deploy images from DockerHub and Azure Container Registries to Azure App Service

To install the Docker extension, press Ctrl+Shift+P, type `ext install`, and then run the Install Extension command to bring up the Marketplace extension list. Next, type **docker** to filter the results, and then select the Docker Support extension, as depicted in Figure 4-23.

![View of the Docker extension for VS Code.](./media/image20.png)

**Figure 4-23**. Installing the Docker Extension in Visual Studio Code

### Step 2: Create a DockerFile related to an existing image (plain OS or dev environments like .NET Core, Node.js, and Ruby)

> You will need a `DockerFile` per custom image to be built and per container to be deployed, therefore, if your app is made up of a single custom service, you will need a single `DockerFile`. But, if your app is composed of multiple services (as in a microservices architecture), you'll need one `Dockerfile` per service.

Reviewer: ```suggestion
You will need a `DockerFile` per custom image to be built and per container to be deployed. If your app is made up of a single custom service, you will need a single `DockerFile`. But if your app is composed of multiple services (as in a microservices architecture), you'll need one `Dockerfile` per service.
```

The `DockerFile` is commonly placed in the root folder of your app or service and contains the required commands so that Docker knows how to set up and run that app or service. You can create your `DockerFile` and add it to your project along with your code (node.js, .NET Core, etc.), or, if you are new to the environment, take a look at the following Tip.

> [!TIP]
>

> > You can use the Docker extension to guide you when using the `Dockerfile` and `docker-compose.yml` files related to your Docker containers. Eventually, you will probably write this kind of files without this tools, but using the Docker extension is a good starting point that will accelerate your learning curve.

Reviewer: ```suggestion
> You can use the Docker extension to guide you when using the `Dockerfile` and `docker-compose.yml` files related to your Docker containers. Eventually, you will probably write these kinds of files without this tool, but using the Docker extension is a good starting point that will accelerate your learning curve.
```

> In figure 4-24, you can see how a docker-compose file is added by using the Docker Extension for VS Code.

Reviewer: ```suggestion
In Figure 4-24, you can see how a docker-compose file is added by using the Docker Extension for VS Code.
```

![Console view of Docker extension for VS Code.](./media/image24.png)

**Figure 4-24**. Docker files added using the **Add Docker files to Workspace command**

When you add a DockerFile, you specify what base Docker image you’ll be using (like using `FROM microsoft/aspnetcore`). You will usually build your custom image on top of a base image that you get from any official repository at the [Docker Hub registry](https://hub.docker.com/) (like an [image for .NET Core](https://hub.docker.com/r/microsoft/dotnet/) or the one [for Node.js](https://hub.docker.com/_/node/)).

***Use an existing official Docker image***

Using an official repository of a language stack with a version number ensures that the same language features are available on all machines (including development, testing, and production).

> Following is a sample DockerFile for a .NET Core container:

Reviewer: ```suggestion
The following is a sample DockerFile for a .NET Core container:
```

In the DockerFile, you can also instruct Docker to listen to the TCP port that you'll use at runtime (such as port 80).

You can specify additional configuration settings in the Dockerfile, depending on the language and framework you're using. For instance, the `ENTRYPOINT` line with `["dotnet", "MySingleContainerWebApp.dll"]` tells Docker to run a .NET Core application. If you're using the SDK and the .NET Core CLI (`dotnet CLI`) to build and run the .NET application, this setting would be different. The key point here is that the ENTRYPOINT line and other settings depend on the language and platform you choose for your application.

> [!INFORMATION]
>
> For more information about building Docker images for .NET Core applications, go to <https://docs.microsoft.com/dotnet/core/docker/building-net-docker-images>.
>
> To learn more about building your own images, go to <https://docs.docker.com/engine/tutorials/dockerimages/>.

**Use multi-arch image repositories**

> A single image name in a repo can contain platform variants, such as a Linux image and a Windows image. This feature allows vendors like Microsoft (base image creators) to create a single repo to cover multiple platforms (that is Linux and Windows). For example, the [microsoft/aspnetcore](https://hub.docker.com/r/microsoft/aspnetcore/) repository available in the Docker Hub registry provides support for Linux and Windows Nano Server by using the same image name.

Reviewer: ```suggestion
A single image name in a repo can contain platform variants, such as a Linux image and a Windows image. This feature allows vendors like Microsoft (base image creators) to create a single repo to cover multiple platforms (that is, Linux and Windows). For example, the [microsoft/aspnetcore](https://hub.docker.com/r/microsoft/aspnetcore/) repository available in the Docker Hub registry provides support for Linux and Windows Nano Server by using the same image name.
```

> When pulling the [microsoft/aspnetcore](https://hub.docker.com/r/microsoft/aspnetcore/) image from a Windows host it pulls the Windows variant, whereas pulling the same image name from a Linux host pulls the Linux variant.

Reviewer: ```suggestion
Pulling the [microsoft/aspnetcore](https://hub.docker.com/r/microsoft/aspnetcore/) image from a Windows host pulls the Windows variant, whereas pulling the same image name from a Linux host pulls the Linux variant.
```

***Create your base image from scratch***

You can create your own Docker base image from scratch as explained in this [article](https://docs.docker.com/engine/userguide/eng-image/baseimages/) from Docker. This scenario is probably not the best for you if you are just starting with Docker, but if you want to set the specific bits of your own base image, you can do it.

### Step 3: Create your custom Docker images embedding your service in it

For each custom service that comprises your app, you'll need to create a related image. If your app is made up of a single service or web app, you'll need just a single image.

> [!NOTE]
>

> > When taking into account the "outer-loop DevOps workflow", the images will be created by an automated build process whenever you push your source code to a Git repository (Continuous Integration) so the images will be created in that global environment from your source code.

Reviewer: ```suggestion
> When taking into account the "outer-loop DevOps workflow", the images will be created by an automated build process whenever you push your source code to a Git repository (Continuous Integration), so the images will be created in that global environment from your source code.
```

>

> > But, before we consider going to that outer-loop route, we need to ensure that the Docker application is actually working properly so that they don't push code that might not work properly to the source control system (Git, etc.).

Reviewer: ```suggestion
> But before we consider going to that outer-loop route, we need to ensure that the Docker application is actually working properly so that they don't push code that might not work properly to the source control system (Git, etc.).
```

**Figure 4-25**. Running docker build

Optionally, instead of directly running `docker build` from the project folder, you first can generate a deployable folder with the .NET libraries needed by using the run `dotnet publish` command, and then run `docker build`.

This example creates a Docker image with the name `cesardl/netcore-webapi-microservice-docker:first` (`:first` is a tag, like a specific version). You can take this step for each custom image you need to create for your composed Docker application with several containers.

You can find the existing images in your local repository (your development machine) by using the docker images command, as illustrated in Figure 4-26.

![Console output from command docker images, showing existing images.](./media/image26.png)

**Figure 4-26**. Viewing existing images using docker images

### Step 4: Define your services in docker-compose.yml when building a composed Docker app with multiple services

With the `docker-compose.yml` file, you can define a set of related services to be deployed as a composed application with the deployment commands explained in the next step section.

> Create that file in your main or root solution folder; it should have a content similar to the shown in this `docker-compose.yml` file:

Reviewer: ```suggestion
Create that file in your main or root solution folder; it should have content similar to that shown in this `docker-compose.yml` file:
```

You can run the Docker image by using the docker run command, as shown here:

```console
docker run -t -d -p 80:5000 cesardl/netcore-webapi-microservice-docker:first
```

> For this particular deployment, we'll be redirecting requests sent to port 80 to the internal port 5000. Now, the application is listening on the external port 80 at the host level.

Reviewer: ```suggestion
For this particular deployment, we'll be redirecting requests sent to port 80 to the internal port 5000. Now the application is listening on the external port 80 at the host level.
```
